src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -409,10 +431,24 @@ public UnboundLambdaState(Binder binder, UnboundLambda unboundLambdaOpt)
         public void SetUnboundLambda(UnboundLambda unbound)
         {
             Debug.Assert(unbound != null);
-            Debug.Assert(_unboundLambda == null);
+            Debug.Assert(_unboundLambda == null || (object)_unboundLambda == unbound);
             _unboundLambda = unbound;
         }
 
+        protected abstract UnboundLambdaState WithCachingCore(bool includeCache);
+
+        internal UnboundLambdaState WithCaching(bool includeCache)
+        {
+            if ((_bindingCache == null) != includeCache)
```

Comment:
I feel this is more clear as `(_bindingCache != null) == includeCache`. Where the LHS of the `==` means "currently has a binding cache" and the RHS means "whether or not I want a binding cache".

Reply:
Minor point though, no need to block anything over it.

Reply:
I'm on the fence - both seem about the same to me. I'll leave as is for now.

---
In reply to: [360737717](https://github.com/dotnet/roslyn/pull/40534#discussion_r360737717) [](ancestors = 360737717)